clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -394,14 +392,14 @@ public void failUnsentRequests(Node node, RuntimeException e) {
     private boolean trySend(long now) {
         // send any requests that can be sent now
         boolean requestsSent = false;
-        for (Map.Entry<Node, List<ClientRequest>> requestEntry: unsent.entrySet()) {
-            Node node = requestEntry.getKey();
-            Iterator<ClientRequest> iterator = requestEntry.getValue().iterator();
-            while (iterator.hasNext()) {
-                ClientRequest request = iterator.next();
-                if (client.ready(node, now)) {
+        Set<Node> nodes = unsent.getNodes();
+        for (Node node: nodes) {
+            if (client.ready(node, now)) {
+                // Remove entry before invoking request callback to avoid callbacks handling
+                // coordinator failures traversing the unsent list again.
+                List<ClientRequest> requests = unsent.remove(node);
```

Comment:
Hmm, doesn't this change mean that ordering can be changed? If another thread enqueues a request right after this `remove` call and `client.ready` returns false, we add back the previous requests after the new requests?